packages/connect-core/src/transform-iterable-story.spec.ts

```diff
@@ -131,102 +189,39 @@ describe("full story", function () {
-          {}
-        )
-        .pipeThrough(
-          streamTransformCompress(compressionReverse, writeMaxBytes, 0),
-          {}
-        )
-        .pipeThrough(streamTransformJoin(writeMaxBytes), {});
-
-      const responseReader = createReadableByteStream(goldenBytes)
-        .pipeThrough(streamTransformSplit(readMaxBytes), {})
-        .pipeThrough(
-          streamTransformDecompress(compressionReverse, readMaxBytes),
-          {}
-        )
-        .pipeThrough(
-          streamTransformParse(serialization, endFlag, endSerialization),
-          {}
-        )
-        .getReader();
-
-      const streamingConn = {
-        async send(message: string): Promise<void> {
-          await requestWriter.ready;
-          await requestWriter.write({
-            end: false,
-            value: message,
-          });
-        },
-        async close(): Promise<void> {
-          await requestWriter.ready;
-          await requestWriter.write({
-            end: true,
-            value: "end",
-          });
-          await requestWriter.close();
-        },
-        async read(): Promise<ReadableStreamReadResult<string>> {
-          const r = await responseReader.read();
-          if (r.done) {
-            return {
-              done: true,
-              value: undefined,
-            };
-          }
-          if (r.value.end) {
-            return {
-              done: true,
-              value: undefined,
-            };
-          }
-          return {
-            done: false,
-            value: r.value.value,
-          };
-        },
-      };
-
-      const writtenBytesPromise = readAllBytes(rawRequestStream);
+      const writer = new Writer<Payload<string, "end">>();
+      const writerIt = transformAsyncIterable(
+        writer,
+        transformSerialize(serialization, endFlag, endSerialization),
+        transformCompress(compressionReverse, writeMaxBytes, 0),
+        transformJoin(writeMaxBytes)
+      );
 
-      await streamingConn.send("alpha");
-      await streamingConn.send("beta");
-      await streamingConn.close();
+      const reader = transformAsyncIterable(
+        new Reader(writerIt),
+        transformSplit(readMaxBytes),
+        transformDecompress(compressionReverse, readMaxBytes),
+        transformParse(serialization, endFlag, endSerialization)
+      );
 
-      const alpha = await streamingConn.read();
-      expect(alpha).toEqual({
-        done: false,
-        value: "alpha",
-      });
-      const beta = await streamingConn.read();
-      expect(beta).toEqual({
-        done: false,
-        value: "beta",
-      });
-      const end = await streamingConn.read();
-      expect(end).toEqual({
-        done: true,
-        value: undefined,
-      });
+      writer.send({ value: "alpha", end: false });
+      writer.send({ value: "beta", end: false });
+      writer.send({ value: "gamma", end: false });
+      writer.send({ value: "delta", end: false });
+      writer.close();
```

Comment:
One thing we have to take care of here: writer.send() should return a promise, that only resolves when the value has been consumed by the other side, the reader. 

Imagine a client that sends gigabytes of data. It needs to wait for the data to actually leave the local buffers.